lib/std/mem.zig

```diff
@@ -635,13 +635,79 @@ test "lessThan" {
 /// Compares two slices and returns whether they are equal.
 pub fn eql(comptime T: type, a: []const T, b: []const T) bool {
     if (a.len != b.len) return false;
-    if (a.ptr == b.ptr) return true;
+    if (a.len == 0 or a.ptr == b.ptr) return true;
+    if (@sizeOf(T) == 0) return true;
+
+    // No vectorization in stage2 x86_64 or x86
+    // https://github.com/ziglang/zig/issues/17748
+    if (builtin.zig_backend != .stage2_x86_64 and builtin.zig_backend != .stage2_x86) {
+        if (std.meta.hasUniqueRepresentation(T)) return eqlBytes(std.mem.sliceAsBytes(a), std.mem.sliceAsBytes(b));
+    }
+
     for (a, b) |a_elem, b_elem| {
         if (a_elem != b_elem) return false;
     }
     return true;
 }
 
+/// std.mem.eql heavily optimized for slices of bytes.
+pub fn eqlBytes(a: []const u8, b: []const u8) bool {
+    if (a.len != b.len) return false;
+    if (a.len == 0 or a.ptr == b.ptr) return true;
```

Comment:
For my own curiosity, I'm wondering how you judged that this optimization was worth keeping?  On the one hand, it's very cheap. On the other, I want to say it'll rarely be true.

Reply:
Which one? You selected 3 different ones.

Reply:
Both the `a.len == 0` and the `a.ptr == b.ptr`

Reply:
Sorry, I accidently started a review last time, missclicked.

TLDR again:
It costs virtually nothing, a mere 2 cmp instructions.
Prevents large cost from happening when it can be exited within the first couple of cycles.
I would not call `a.len == 0` to be rare, it's definitely something that can happen. `a.ptr == b.ptr` is something that also can happen and is a useful check to make. It could be some sort of overlap happening, who knows what. 

There's just no point *not* adding them. They cost virtually nothing.

Reply:
You can't measure the cost of something just by analyzing the number of instructions involved - but even if you could, it's not just the `cmp`, as there's also a conditional branch involved here, which is the real source of potential trouble.

In this case, the optimizer may make the wrong choice for how to construct this branch (does "taking the branch" enter the `if`, or skip it?) and lead to a non-trivial cycle count, particularly if mispredicted.

If Zig had likely/unlikely hints, the cost of this code would indeed be very little, as a branch not taken costs a tiny fraction of a cycle. Currently, we do not have such hints. Personally, I would suggest keeping this check, but putting a comment referencing the accepted (but not yet implemented) proposal for an `@expect` builtin.

Reply:
Of course I understand that the pure instructions are not the only factor. I might have no mentioned it in my initial reply, but I meant to say there is no actual performance difference. While I understand what you mean about the optimizer and branching, in the end looking at the raw numbers shows no noticeable performance change.

Even then, I just sort of disagree. Looking at the instructions is one of the only ways to tell if the cost will truly matter. The optimizer has finished working on that specific code. There's no branch prediction or missing happening, or question about it. If I had said I was looking at the Air or llvm IR, then obviously this is in a non-optimized state. But objdumping the binary itself does yield the final instructions, and I believe is an accurate way to quickly measure cost.

Reply:
While an expected keyword would be massive for the language, and something I desperately desire it doesn't exist yet. My make do is simply doing raw empirical measurements. Looking at the generated assembly on many different targets, in different modes. 

Ive attempted to mimic the expected builtin by making a separate setCold(true) function for returning, and using that. Doesn't affect code generation, because it's already generating the most optimal code.